File: app-android/app/src/main/java/de/tutao/tutanota/push/SseStorage.java
package de.tutao.tutanota.push;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;

import org.apache.commons.io.IOUtils;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.KeyStoreException;
import java.security.UnrecoverableEntryException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import de.tutao.tutanota.AndroidKeyStoreFacade;
import de.tutao.tutanota.CryptoError;
import de.tutao.tutanota.Utils;
import de.tutao.tutanota.alarms.AlarmNotification;
import de.tutao.tutanota.data.AppDatabase;
import de.tutao.tutanota.data.PushIdentifierKey;
import de.tutao.tutanota.data.SseInfo;
import de.tutao.tutanota.data.User;


public class SseStorage {
	private static final String LAST_PROCESSED_NOTIFICATION_ID = "lastProcessedNotificationId";
	private static final String LAST_MISSED_NOTIFICATION_CHECK_TIME = "'lastMissedNotificationCheckTime'";
	private static final String DEVICE_IDENTIFIER = "deviceIdentifier";
	private static final String SSE_ORIGIN = "sseOrigin";
	public static final String CONNECT_TIMEOUT_SEC = "connectTimeoutSec";

	private final AppDatabase db;
	private final AndroidKeyStoreFacade keyStoreFacade;

	public SseStorage(AppDatabase db, AndroidKeyStoreFacade keyStoreFacade) {
		this.db = db;
		this.keyStoreFacade = keyStoreFacade;
	}

	@Nullable
	public String getPushIdentifier() {
		return db.keyValueDao().getString(DEVICE_IDENTIFIER);
	}


	public void storePushIdentifier(String identifier, String sseOrigin) {
		db.keyValueDao().putString(DEVICE_IDENTIFIER, identifier);
		db.keyValueDao().putString(SSE_ORIGIN, sseOrigin);
	}

	public void clear() {
		setLastMissedNotificationCheckTime(null);

		db.userInfoDao().clear();
		db.getAlarmInfoDao().clear();
	}

	public void storePushIdentifierSessionKey(String userId, String pushIdentiferId,
											  String pushIdentifierSessionKeyB64) throws KeyStoreException, CryptoError {
		byte[] deviceEncSessionKey = this.keyStoreFacade.encryptKey(Utils.base64ToBytes(pushIdentifierSessionKeyB64));
		this.db.userInfoDao().insertPushIdentifierKey(new PushIdentifierKey(pushIdentiferId, deviceEncSessionKey));
		this.db.userInfoDao().insertUser(new User(userId));
	}


	public byte[] getPushIdentifierSessionKey(String pushIdentiferId) throws UnrecoverableEntryException, KeyStoreException, CryptoError {
		PushIdentifierKey userInfo = this.db.userInfoDao().getPushIdentifierKey(pushIdentiferId);
		if (userInfo == null) {
			return null;
		}
		return this.keyStoreFacade.decryptKey(userInfo.getDeviceEncPushIdentifierKey());
	}

	public LiveData<List<User>> observeUsers() {
		return this.db.userInfoDao().observeUsers();
	}

	public List<AlarmNotification> readAlarmNotifications() {
		return db.getAlarmInfoDao().getAlarmNotifications();
	}

	public void insertAlarmNotification(AlarmNotification alarmNotification) {
		db.getAlarmInfoDao().insertAlarmNotification(alarmNotification);
	}

	public void deleteAlarmNotification(String alarmIdetifier) {
		db.getAlarmInfoDao().deleteAlarmNotification(alarmIdetifier);
	}

	@Nullable
	@WorkerThread
	public String getLastProcessedNotificationId() {
		return db.keyValueDao().getString(LAST_PROCESSED_NOTIFICATION_ID);
	}

	@WorkerThread
	public void setLastProcessedNotificationId(String id) {
		db.keyValueDao().putString(LAST_PROCESSED_NOTIFICATION_ID, id);
	}

	@Nullable
	@WorkerThread
	public Date getLastMissedNotificationCheckTime() {
		long value = db.keyValueDao().getLong(LAST_MISSED_NOTIFICATION_CHECK_TIME);
		if (value == 0) {
			return null;
		}
		return new Date(value);
	}

	@WorkerThread
	public void setLastMissedNotificationCheckTime(@Nullable Date date) {
		db.keyValueDao().putLong(LAST_MISSED_NOTIFICATION_CHECK_TIME, date == null ? 0L : date.getTime());
	}

	public String getSseOrigin() {
		return this.db.keyValueDao().getString(SSE_ORIGIN);
	}

	public long getConnectTimeoutInSeconds() {
		return db.keyValueDao().getLong(CONNECT_TIMEOUT_SEC);
	}

	public void setConnectTimeoutInSeconds(long connectTimeout) {
		db.keyValueDao().putLong(CONNECT_TIMEOUT_SEC, connectTimeout);
	}

	public void removeUser(String userId) {
		this.db.userInfoDao().deleteUser(userId);
	}

	public List<User> getUsers() {
		return this.db.userInfoDao().getUsers();
	}
}

File: app-android/app/src/main/java/de/tutao/tutanota/push/SseStorage.kt
package de.tutao.tutanota.push

import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import de.tutao.tutanota.AndroidKeyStoreFacade
import de.tutao.tutanota.CryptoError
import de.tutao.tutanota.alarms.AlarmNotificationEntity
import de.tutao.tutanota.data.AppDatabase
import de.tutao.tutanota.data.PushIdentifierKey
import de.tutao.tutanota.data.User
import java.security.KeyStoreException
import java.security.UnrecoverableEntryException
import java.util.*

class SseStorage(
		private val db: AppDatabase,
		private val keyStoreFacade: AndroidKeyStoreFacade,
) {
	fun getPushIdentifier() = db.keyValueDao().getString(DEVICE_IDENTIFIER)

	fun storePushIdentifier(identifier: String, sseOrigin: String) {
		db.keyValueDao().putString(DEVICE_IDENTIFIER, identifier)
		db.keyValueDao().putString(SSE_ORIGIN, sseOrigin)
	}

	fun clear() {
		setLastMissedNotificationCheckTime(null)
		db.userInfoDao().clear()
		db.alarmInfoDao().clear()
	}

	@Throws(KeyStoreException::class, CryptoError::class)
	fun storePushIdentifierSessionKey(
			userId: String,
			pushIdentifierId: String,
			pushIdentifierSessionKey: ByteArray,
	) {
		val deviceEncSessionKey = keyStoreFacade.encryptKey(pushIdentifierSessionKey)
		db.userInfoDao().insertPushIdentifierKey(PushIdentifierKey(pushIdentifierId, deviceEncSessionKey))
		db.userInfoDao().insertUser(User(userId))
	}

	@Throws(UnrecoverableEntryException::class, KeyStoreException::class, CryptoError::class)
	fun getPushIdentifierSessionKey(pushIdentifierId: String): ByteArray? {
		val userInfo = db.userInfoDao().getPushIdentifierKey(pushIdentifierId) ?: return null
		return keyStoreFacade.decryptKey(userInfo.deviceEncPushIdentifierKey!!)
	}

	fun observeUsers(): LiveData<List<User>> {
		return db.userInfoDao().observeUsers()
	}

	fun readAlarmNotifications(): List<AlarmNotificationEntity> {
		return db.alarmInfoDao().alarmNotifications
	}

	fun insertAlarmNotification(alarmNotification: AlarmNotificationEntity) {
		db.alarmInfoDao().insertAlarmNotification(alarmNotification)
	}

	fun deleteAlarmNotification(alarmIdentifier: String) {
		db.alarmInfoDao().deleteAlarmNotification(alarmIdentifier)
	}

	@WorkerThread
	fun getLastProcessedNotificationId() = db.keyValueDao().getString(LAST_PROCESSED_NOTIFICATION_ID)

	@WorkerThread
	fun setLastProcessedNotificationId(id: String) = db.keyValueDao().putString(LAST_PROCESSED_NOTIFICATION_ID, id)

	@WorkerThread
	fun getLastMissedNotificationCheckTime(): Date? {
		val value = db.keyValueDao().getLong(LAST_MISSED_NOTIFICATION_CHECK_TIME)
		return if (value == 0L) {
			null
		} else Date(value)
	}

	@WorkerThread
	fun setLastMissedNotificationCheckTime(date: Date?) = db.keyValueDao().putLong(LAST_MISSED_NOTIFICATION_CHECK_TIME, date?.time
			?: 0L)

	fun getSseOrigin() = db.keyValueDao().getString(SSE_ORIGIN)

	fun getConnectTimeoutInSeconds() = db.keyValueDao().getLong(CONNECT_TIMEOUT_SEC)

	fun setConnectTimeoutInSeconds(timeout: Long) = db.keyValueDao().putLong(CONNECT_TIMEOUT_SEC, timeout)

	fun removeUser(userId: String) {
		db.userInfoDao().deleteUser(userId)
	}

	fun getUsers(): List<User> = db.userInfoDao().users

	companion object {
		private const val LAST_PROCESSED_NOTIFICATION_ID = "lastProcessedNotificationId"
		private const val LAST_MISSED_NOTIFICATION_CHECK_TIME = "'lastMissedNotificationCheckTime'"
		private const val DEVICE_IDENTIFIER = "deviceIdentifier"
		private const val SSE_ORIGIN = "sseOrigin"
		const val CONNECT_TIMEOUT_SEC = "connectTimeoutSec"
	}
}